crates/aptos-crypto/src/bls12381.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides API for the BLS12-381 signature library (https://github.com/supranational/blst).
+
+use crate::{
+    hash::CryptoHash, signing_message, CryptoMaterialError, Length, PrivateKey, PublicKey,
+    Signature, SigningKey, Uniform, ValidCryptoMaterial, ValidCryptoMaterialStringExt,
+    VerifyingKey,
+};
+use anyhow::{anyhow, Result};
+use aptos_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use blst::{blst_scalar, BLST_ERROR};
+use rand_core::{OsRng, RngCore};
+use serde::Serialize;
+use std::convert::TryFrom;
+
+const DST: &[u8] = b"BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";
+
+#[derive(Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// A BLS12381 public key
+pub struct BLS12381PublicKey {
+    pubkey: blst::min_sig::PublicKey,
+}
+
+#[derive(SerializeKey, DeserializeKey, SilentDebug, SilentDisplay)]
+/// A BLS12381 private key
+pub struct BLS12381PrivateKey {
+    privkey: blst::min_sig::SecretKey,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// A BLS12381 signature
+pub struct BLS12381Signature {
+    sig: blst::min_sig::Signature,
+}
+
+impl BLS12381PublicKey {
+    /// The length of the BLS12381PublicKey
+    pub const LENGTH: usize = 96;
+
+    /// Serialize a BLS12381PublicKey.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.pubkey.to_bytes()
+    }
+
+    /// Validate the public key.
+    pub fn validate(&self) -> Result<()> {
+        self.pubkey.validate().map_err(|e| anyhow!("{:?}", e))
+    }
+}
+
+impl BLS12381PrivateKey {
+    /// The length of the BLS12381PrivateKey
+    pub const LENGTH: usize = 32;
+
+    /// Serialize a BLS12381PrivateKey.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.privkey.to_bytes()
+    }
+}
+
+impl BLS12381Signature {
+    /// The length of the BLS12381Signature
+    pub const LENGTH: usize = 48;
+
+    /// Serialize a BLS12381Signature.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.sig.to_bytes()
+    }
+
+    /// Validate the signature.
+    pub fn validate(&self) -> Result<()> {
+        self.sig.validate(true).map_err(|e| anyhow!("{:?}", e))
+    }
+
+    /// Aggregate multiple signatures, assume individual signature are validated.
```

Comment:
This comments says you assume individual signature shares are validated, but in practice we will NOT want to do that, since it will make aggregation painfully slow in the optimistic case.